programs/ssl/ssl_client2.c

```diff
@@ -419,10 +428,44 @@ struct options
     int dgram_packing;          /* allow/forbid datagram packing            */
     int extended_ms;            /* negotiate extended master secret?        */
     int etm;                    /* negotiate encrypt then mac?              */
+    int eap_tls;                /* derive EAP-TLS keying material?          */
 } opt;
 
 int query_config( const char *config );
 
+#if defined(MBEDTLS_SSL_EXPORT_KEYS)
+typedef struct eap_tls_keys
+{
+    unsigned char master_secret[48];
+    unsigned char randbytes[64];
+    mbedtls_ssl_tls_prf *tls_prf;
+} eap_tls_keys;
+
+static int eap_tls_key_derivation ( void *p_expkey,
```

Comment:
Minor: unnecessary space before parenthesis.

Reply:
I agree it requires testing, but we don't have any test infrastructure for comparing buffers. As mentioned also in #1813 
I don't think this should be part of the scope of this PR though

Reply:
This function is EAP TLS key derivation only in its name, it exports the key material and the derivation happens later.

I agree that we shouldn't burden the library with EAP TLS key derivation testing. (But I still think that having the key derivation here is useful example of the key export feature.)

This PR extends the key export feature and I was suggesting to test only that. We usually test high level features like this by printing out something if they work and look for it in the logs when we add a test in `ssl-opt.sh`. (I only mentioned `all.sh` because I didn't realise that `MBEDTLS_SSL_EXPORT_KEYS` is in the default configuration.) I suggest to use something predictable for reliable testing (not the premaster secret or the random values).

Reply:
Exactly, this PR extends the key export feature, not the PRF feature. ( until now).
The key export exports the premaster secret and random values ( among other things), and I don't see a way how to test that.
As for testing the `tls_prf` I guess I can change the applications to use `dummy_keys` for tls_prf, and print the values for testing.

Reply:
I see that with our current infrastructure it can't be tested fully, I just would like to test it to an extent we can. Namely that it is called and at least some data exported is correct. We can't test all the exported data, but I think it is valuable to test the little we can.

I think now that we are exposing `tls_prf` in the public API, we can test it in unit test and we shouldn't test it through this sample program.

Reply:
I see. I'll add a unit test

Reply:
We could print here `maclen` and/or `keylen` and/or `ivlen` and add a test to `ssl-opt.sh` to make sure that the callback is called and that it does print the right values.

Reply:
It is doable, but I don't like the result.. We should be using the SSL DEBUG functionality for that, but it's not available in this callback.
otherwise, we will always have something like:
~~~
  . Performing the SSL/TLS handshake...
exported maclen is 0
exported keylen is 32
exported ivlen is 12
 ok
~~
no matter what debug level we enabled.
Any suggestion?

Reply:
I can think of two solutions:
- Let's export these values too and print them before printing the derived EAP key material. This way it will be conceptually a stub/simulation of application logic and not some debug information. Consequently it is perfectly fine to  print it out unconditionally.
- Or just make the printing conditional on the debug level.

I find the second option simpler and more straightforward and I would suggest implementing that.

Reply:
I agree with the second solution. Thanks!